docs: update NG0506

arturovt opened the pull request:
Provide a thorough explanation of why an application may remain unstable and why developers should consider not running asynchronous code on the server outside of the Angular zone, as this may lead to memory leaks.

arturovt committed:
docs: update NG0506

Provide a thorough explanation of why an application may remain unstable and
why developers should consider not running asynchronous code on the server outside
of the Angular zone, as this may lead to memory leaks.

JeanMeche commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -2,19 +2,19 @@
-Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame` etc.
-If one of these functions is called in the initialization phase of the app or the bootstrapped components, the application will remain unstable.
+Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.
+If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.
```
Should we make these sentences "ready" for zoneless ?

JeanMeche commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -2,19 +2,19 @@
 This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
```
> The `ApplicationRef.isStable` API uses NgZone 
In v18 it uses the internal `PendingTasks` service.

arturovt replied:
Made updates.

pkozlowski-opensource commented:
I would expect mention of the `afterRender` / `afterNextRender` hooks in this docs as a place to locate logic that needs to run in a browser (as an alternative to running code outside of zones)

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -2,19 +2,23 @@
 
 This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks. In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`.
```
I think we can rename the title to "Application remains unstable", i.e. drop "NgZone", since the problem may technically happen in Zoneless mode too. I'd also propose to drop the "In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`." sentence - it's an internal implementation detail (may not have a lot of value to developers debugging a problem). We can mention though that https://angular.dev/api/core/ExperimentalPendingTasks also contribute to the application stability.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
 ```
 
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during server-side rendering, or if they should only be triggered in the browser. Running timers on the server may result in delayed response times, as the Angular framework waits for the application to stabilize.
````
This error is thrown in the browser during the hydration process. I'm not sure if mentioning server part of it would be relevant in this guide.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during server-side rendering, or if they should only be triggered in the browser. Running timers on the server may result in delayed response times, as the Angular framework waits for the application to stabilize.
+
+Invoking those functions outside of the Angular zone on the server is considered a "dirty" hack because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.
```
Similar to the previous comment - this may not be relevant for this guide (we can find a different location in the docs for this info).

If we end up leaving this paragraph, I'd like to propose an update:
```suggestion
Invoking those functions outside of the Angular zone on the server is not recommended because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+
+### Using `afterNextRender`
+
+You may also resort to using the [`afterNextRender`](api/core/afterNextRender) function, which could provide a cleaner method for executing browser-only code outside of the Angular zone. `afterNextRender` callbacks are automatically invoked outside of the Angular zone and are executed only once, whereas `afterRender` will invoke the provided callback after each render:
```
This behavior has changed recently, the callbacks are no longer invoke outside of NgZone explicitly: https://github.com/angular/angular/commit/7e89753eeff24f52d39fef92600293bf1700cd1b

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+}
+```
+
+When piping streams with `debounceTime` and specifying a timeout to debounce generated events, it's important to note that this practice may contribute to application instability, particularly during server-side rendering. It's advisable to check the current platform and adjust the timeout settings accordingly on the server side:
````
This also seems to be related to server vs client (see comments above).

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
-## Debugging the error
+### Third-party libraries
+
+Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some of these libraries may schedule tasks during their initialization logic, which can prevent Angular from stabilizing properly. For instance, consider LogRocket, a session replay platform, which may set up a timer using `setTimeout` to send network information. The application will remain unstable until the ...
```
I'd propose to avoid referencing a specific library, just mention that some third-party libraries may have patterns that would delay application stability. The recommendation might be the same: try running relevant library code outside of zone.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+}
+```
+
+### Running code after the hydration is done
+
+You may also delay the code that sets up asynchronous tasks until the hydration process is complete to prevent them from delaying the actual hydration process:
````
That's a good point 👍 I'd propose rephrasing it to say that you can run a code once an application becomes stable (vs "when hydration is done")

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+You may also delay the code that sets up asynchronous tasks until the hydration process is complete to prevent them from delaying the actual hydration process:
+
+```typescript
+import LogRocket from 'logrocket';
````
Similar to one of the previous comments, let's avoid using a specific library.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+
+## Zoneless applications
+
+Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```
```suggestion
[Zoneless applications](./guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+
+In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
+
+Developers may also explicitly contribute to indicating the application's stability by using the experimental `PendingTasks` service.
```
It'd be great to have a link to API docs for the `ExperimentalPendingTasks` symbol here.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+
+Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
+
+In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
```
```suggestion
In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
+    });
+  }
+}
+```
+
+### Running code after the hydration is done
+
+You may also delay the code that sets up asynchronous tasks until the hydration process is complete to prevent them from delaying the actual hydration process:
+
+```typescript
+import LogRocket from 'logrocket';
+
+class SimpleComponent {
+  constructor() {
+    const platformId = inject(PLATFORM_ID);
+
+    if (isPlatformBrowser(platformId)) {
+      const applicationRef = inject(ApplicationRef);
+
+      applicationRef.isStable.pipe( first((isStable) => isStable) ).subscribe(() => {
+        // Note that we don't need to use `runOutsideAngular` because `isStable`
+        // emits events outside of the Angular zone when it's truthy (falsy values
+        // are emitted inside the Angular zone).
+        LogRocket.init(...);
+      });
+    }
+  }
+}
+```
+
+## Zoneless applications
+
+Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
+
+In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
+
+Developers may also explicitly contribute to indicating the application's stability by using the experimental `PendingTasks` service.
 
-Verify that you don't have any long standing microtask or macrotasks.
+In conclusion, it's crucial to monitor all asynchronous tasks scheduled within the application and determine whether they are necessary. Additionally, verifying whether third-party libraries use any asynchronous APIs internally is important, as they may contribute to application instability.
````
I'd propose dropping this paragraph, this would mimic the table of contents on a page and info from the first paragraphs.

AndrewKushnir reviewed (commented):
@arturovt thanks for improving this guide 👍

It looks like some of the recommendations may not be applicable (but some of them might potentially be moved to a different guide), since this error only happens on the client.

arturovt commented:
@AndrewKushnir I have made updates based on all of your feedback. I removed the parts mentioning server-side rendering. Initially, I included server-side rendering information, but I realized that this warning is specific to browsers only. Additionally, I removed the mention of `afterNextRender` because it is not invoked outside of the zone by default anymore and does not add any value to this guide.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -1,20 +1,24 @@
+# Application remains unstable
 
-This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
+This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period.
```
```suggestion
This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period of time (over 10 seconds).
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -1,20 +1,24 @@
+- in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed
 
-This warning is displayed when the `ApplicationRef.isStable()` doesn't emit `true` within 10 seconds. If this is intentional and your application becomes stable later, you can ignore this warning.
+This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning. However, it is not recommended to maintain long-running tasks in the background that prevent the application from stabilizing.
```
I'd propose removing the last sentence, since it might be intentional (as mentioned earlier in this paragraph):
```suggestion
This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -1,20 +1,24 @@
+This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
```
I think we can rephrase this to drop the mention of NgZone (since it's not used in Zoneless mode):
```suggestion
Typically that happens when there are some pending microtasks or macrotasks on a page.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -1,20 +1,24 @@
 
-Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame` etc.
-If one of these functions is called in the initialization phase of the app or the bootstrapped components, the application will remain unstable.
+Applications that still use zone.js may encounter various factors contributing to instability. These may include pending HTTP requests, timers, or even `fs.readFile` calls, if it's used during server-side rendering on the Node.js platform, as certain Node.js APIs are also partially patched by zone.js.
```
I've removed fs and Node-based content, since it's not relevant on the client (where an error happens):
```suggestion
Applications that use zone.js may have various factors contributing to delays in stability. These may include pending HTTP requests, timers (`setInterval`, `setTimeout`) or some logic that continuously invokes `requestAnimationFrame`.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -1,20 +1,24 @@
+### Macrotasks
+
+Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.
+If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.
```
```suggestion
If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may delay the moment when the application becomes stable.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
````diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
 ```
 
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during the bootstrap process. Running timers during bootstrap may result in delayed hydration cleanup times, as the Angular framework waits for the application to stabilize.
````
I'd propose removing this sentence, it looks like it repeats the content mentioned earlier.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during the bootstrap process. Running timers during bootstrap may result in delayed hydration cleanup times, as the Angular framework waits for the application to stabilize.
+
+If you need to schedule those timers in the browser, they should also be wrapped with `runOutsideAngular` to allow for hydration cleanup as soon as possible:
```
We should also rephrase or revert this sentence.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
+  constructor() {
+    const platformId = inject(PLATFORM_ID);
+
+    if (isPlatformBrowser(platformId)) {
```
Since the error happens on the client only, there is no need to check for `isPlatformBrowser` here. I'd propose keeping this example simple (similar to the current code).

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
+
+### Third-party libraries
+
+Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some third-party libraries may have patterns that could delay application stability. The recommendation remains the same: try executing relevant library code outside of the zone.
```
```suggestion
Some third-party libraries can also produce long-running asynchronous tasks, which may delay application stability. The recommendation is to invoke relevant library code outside of the zone as described above.
```

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
+  constructor() {
+    const platformId = inject(PLATFORM_ID);
+
+    if (isPlatformBrowser(platformId)) {
```
Do we need `isPlatformBrowser` in this example?

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
-## Debugging the error
+## Zoneless applications
+
+[Zoneless applications](/guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTT ...
```
This is an implementation detail, I'd propose removing this paragraph.

AndrewKushnir commented on adev/src/content/reference/errors/NG0506.md:
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
+In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
 
-Verify that you don't have any long standing microtask or macrotasks.
+Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service.
```
```suggestion
Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service. If you use the mentioned APIs in your application, make sure you invoke a function to mark the task as completed.
```

arturovt commented:
@AndrewKushnir made changes.

AndrewKushnir reviewed (approved):
@arturovt thanks for addressing the feedback and improving this page 👍

alxhub commented:
This PR was merged into the repository by commit 2781b20567f7c717e4d8595997cd61dc88ca7cce.
